crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,184 @@
+    fn message(&self) -> String {
+        format!("Missing author into TODO")
+    }
+}
+
+#[violation]
+pub struct MissingLink;
+impl Violation for MissingLink {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("To be implemented")
+    }
+}
+
+#[violation]
+pub struct MissingColonInTodo;
+impl Violation for MissingColonInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing colon in TODO")
+    }
+}
+
+#[violation]
+pub struct MissingTextInTodo;
+impl Violation for MissingTextInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing text in TODO")
+    }
+}
+
+#[violation]
+pub struct InvalidCapitalizationInTodo {
+    pub tag: String,
+}
+impl Violation for InvalidCapitalizationInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let InvalidCapitalizationInTodo { tag } = self;
+        format!("Invalid TODO capitalization: `{tag}` should be `TODO`")
+    }
+}
+
+#[violation]
+pub struct MissingSpaceAfterColonInTodo;
+impl Violation for MissingSpaceAfterColonInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing space after colon in TODO")
+    }
+}
+
+// Capture groups correspond to checking:
+//      1. Tag (used to match against `TODO` in capitalization and spelling, e.g. ToDo and FIXME)
+//      2. Author exists (in parentheses) after tag, but before colon
+//      3. Colon exists after author
+//      4. Space exists after colon
+//      5. Text exists after space
+//
+// We can check if any of these exist in one regex. Capture groups that don't pick anything up
+// evaluate to `None` in Rust, so the capture group index will always correspond to its respective rule
+// whether the token has been found or not.
+//
+// Example:
+// ```python
+// # TODO(evanrittenhouse): This is a completely valid TODO
+// ```
+// will yield [Some("TODO") Some("evanrittenhouse"), Some(":"), Some(" "), Some("This is a completely valid TODO")], whereas
+// ```python
+// # ToDo this is completely wrong
+// ```
+// will yield [Some("ToDo"), None, None, Some(" "), Some("this is completely wrong")]. Note the
+// `Nones` for the colon and space checks
+//
+// Note: Tags taken from https://github.com/orsinium-labs/flake8-todos/blob/master/flake8_todos/_rules.py#L12.
+static TODO_REGEX: Lazy<Regex> = Lazy::new(|| {
+    // TODO BEFORE COMMITTING - <space> should be a nested group inside of <colon>
+    Regex::new(r"^#\s*([tT][oO][dD][oO]|BUG|FIXME|XXX)(\(.*\))?(:)?( )?(.+)?$").unwrap()
+});
+static NUM_CAPTURE_GROUPS: usize = 5usize;
+
+pub fn check_rules(tokens: &[LexResult]) -> Vec<Diagnostic> {
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+
+    for (start, token, end) in tokens.iter().flatten() {
+        let Tok::Comment(comment) = token else {
+            continue;
+        };
+
+        if get_captured_matches(comment).peek().is_some() {
```

Comment:
`get_captured_matches()` and `get_tag_regex_errors()` both call `TODO_REGEX.captures_iter()`, an expensive operation. I'll need to refactor that to only call it once and pass the capture groups around before this gets merged.